### PR TITLE
added whitelist functionality to config.json

### PR DESCRIPTION
**Pull request by mitkotak**
Picks up where #43 left off with a different branch

### CODE REVIEW

**Comment by alexandermichels** [comment]
Could we use this list as our whitelist: https://github.com/cybergis/cybergis-compute-core/blob/v2/configs/jupyter-globus-map.example.json ?

**Comment by alexandermichels** [comment]
> Could we use this list as our whitelist: https://github.com/cybergis/cybergis-compute-core/blob/v2/configs/jupyter-globus-map.example.json ?

Any follow-up on this?

**Comment by mitkotak** [comment]
> > Could we use this list as our whitelist: https://github.com/cybergis/cybergis-compute-core/blob/v2/configs/jupyter-globus-map.example.json ?
> 
> Any follow-up on this?

Done ! Sorry forgot to push it earlier.

**Review by alexandermichels — COMMENTED** [review]
It seems to be working correctly, passed the api tests, and I tested that CyberGISX/hsdocs-dev are both still able to submit jobs, download, and work correctly. 

The error doesn't seem to be very informative though, just saying "invalid token" instead of the actual problem: that the Hub isn't in the whitelist.

```
---------------------------------------------------------------------------
Exception                                 Traceback (most recent call last)
/tmp/ipykernel_110/1070237876.py in <module>
----> 1 cybergis.show_ui()

/opt/conda/lib/python3.9/site-packages/cybergis_compute_client/CyberGISCompute.py in show_ui(self, input_params, defaultJob, defaultDataFolder, defaultRemoteResultFolder, jupyterhubApiToken)
    519         if df is not None:
    520             self.ui.defaultRemoteResultFolder = df if df[0] == '/' else '/' + df
--> 521         self.ui.render()
    522 
    523     def get_latest_created_job(self):

/opt/conda/lib/python3.9/site-packages/cybergis_compute_client/UI.py in render(self)
     78         """
     79         self.init()
---> 80         self.renderComponents()
     81         divider = Markdown('***')
     82         # render main UI

/opt/conda/lib/python3.9/site-packages/cybergis_compute_client/UI.py in renderComponents(self)
    148         self.renderResultLogs()
    149         self.renderDownload()
--> 150         self.renderRecentlySubmittedJobs()
    151         self.renderLoadMore()
    152         self.renderSubmitNew()

/opt/conda/lib/python3.9/site-packages/cybergis_compute_client/UI.py in renderRecentlySubmittedJobs(self)
    515         if self.recently_submitted['output'] is None:
    516             self.recently_submitted['output'] = widgets.Output()
--> 517             jobs = self.compute.client.request('GET', '/user/job', {'jupyterhubApiToken': self.compute.jupyterhubApiToken})
    518         with self.recently_submitted['output']:
    519             display(Markdown('**Recently Submitted Jobs for ' + self.compute.username.split('@')[0] + '**'))

/opt/conda/lib/python3.9/site-packages/cybergis_compute_client/Client.py in request(self, method, uri, body)
     65             if 'messages' in data:
     66                 msg = str(data['messages'])
---> 67             raise Exception('server ' + self.url + uri + ' responded with error "' + data['error'] + msg + '"')
     68 
     69         return data

Exception: server cgjobsup-dev.cigi.illinois.edu:443/user/job responded with error "invalid token"
```

The logs are just reporting a 402 on /user/job

```
10.0.93.4 - - [23/Sep/2022:19:41:47 +0000] "GET /git HTTP/1.1" 200 14416 "-" "-"
10.0.93.4 - - [23/Sep/2022:19:41:47 +0000] "GET /hpc HTTP/1.1" 200 1270 "-" "-"
10.0.93.4 - - [23/Sep/2022:19:41:47 +0000] "GET /user/job HTTP/1.1" 402 25 "-" "-"
```

I also tried to be explicit as possible by manually setting my username and still didn't get back the "Cannot find jupyterhubHost in whitelist" error. 

Interestingly, it seems that we inadvertently had a whitelist anyways. When I tried to submit a job to the production from https://jupyterhub.cuahsi.org/, the UI came up, but when I tried to submit a job I got the following error in the JupyterLab log (not on the notebook UI):

```
---------------------------------------------------------------------------
Exception                                 Traceback (most recent call last)
/opt/conda/lib/python3.9/site-packages/cybergis_compute_client/UI.py in on_click(change)
    607             data = self.get_data()
    608             if data['computing_resource'] != 'local_hpc':
--> 609                 self.jupyter_globus = self.compute.get_user_jupyter_globus()
    610             if self.job['require_upload_data']:
    611                 dataPath = self.uploadData['selector'].selected

/opt/conda/lib/python3.9/site-packages/cybergis_compute_client/CyberGISCompute.py in get_user_jupyter_globus(self)
    553              Job: Latest Job object instance
    554         """
--> 555         return self.client.request(
    556             'GET', '/user/jupyter-globus', {
    557                 "jupyterhubApiToken": self.jupyterhubApiToken})

/opt/conda/lib/python3.9/site-packages/cybergis_compute_client/Client.py in request(self, method, uri, body)
     65             if 'messages' in data:
     66                 msg = str(data['messages'])
---> 67             raise Exception('server ' + self.url + uri + ' responded with error "' + data['error'] + msg + '"')
     68 
     69         return data

Exception: server cgjobsup.cigi.illinois.edu:443/user/jupyter-globus responded with error "unknown host"
```

It's still good to explicitly include though since other routes weren't protected.

## Two Questions:

Great work @mitkotak, but I have two questions before I approve:

1. How difficult would it be to have the SDK report the "Cannot find jupyterhubHost in whitelist" error instead of "invalid token"?
2. Can we add a whitelist route similar to container? https://github.com/cybergis/cybergis-compute-core/blob/v2/server.ts#L425 I think if we did this for the whitelist, we probably only want to expose the keys (the hostnames) and not the values (endpoints, etc.) just in case. This would be invaluable for debugging and would allow us to implement SDK side checks in case the warnings/errors coming from the Core aren't clean enough.

P.S. I recommend making a Hydroshare account https://www.hydroshare.org/ and use their JupyterHub (you'll need to submit a request) for testing since they aren't currently in our whitelist.

File: server.ts [hunk 1]
```diff
@@ -243,6 +244,13 @@ app.get("/user", (req, res) => {
     res.status(402).json({ error: "invalid input", messages: errors });
     return;
   }
+
+  var jupyterGlobus = jupyterGlobusMap[res.locals.host];
+  if (!jupyterGlobus) {
+    res.status(404).json({ error: "Cannot find jupyterhubHost in whitelist" });
+    return;
```
Comment (alexandermichels) [inline server.ts:251]: I don't like that we are repeating this codeblock so much (maybe make it a function and call it?), but for now this will work fine.

**Review by alexandermichels — APPROVED** [review]
Everything seems to be working, great work @mitkotak! The new route makes perfect sense and the user is now getting the following message:

image

**Pull request merged**